pkg/operator/csidriveroperator/driverstarter.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/klog/v2"
+)
+
+const (
+	infraConfigName = "cluster"
+)
+
+// This CSIDriverStarterController starts CSI driver controllers based on the
+// underlying cloud. It does not install anything by itself, only monitors
+// Infrastructure instance and starts individual ControllerManagers for the
+// particular cloud. It produces following Conditions:
+// CSIDriverStarterDegraded - error checking the Infrastructure
+type CSIDriverStarterController struct {
+	operatorClient v1helpers.OperatorClient
+	infraLister    openshiftv1.InfrastructureLister
+	versionGetter  status.VersionGetter
+	targetVersion  string
+	eventRecorder  events.Recorder
+
+	controllers []csiDriverControllerManager
+}
+
+type csiDriverControllerManager struct {
+	name           string
+	operatorConfig csioperatorclient.CSIOperatorConfig
+	platform       configv1.PlatformType
+	mgr            manager.ControllerManager
+	running        bool
+}
+
+func NewCSIDriverStarterController(
+	clients *csoclients.Clients,
+	resyncInterval time.Duration,
+	versionGetter status.VersionGetter,
+	targetVersion string,
+	eventRecorder events.Recorder,
+	driverConfigs []csioperatorclient.CSIOperatorConfig) factory.Controller {
+	c := &CSIDriverStarterController{
+		operatorClient: clients.OperatorClient,
+		infraLister:    clients.ConfigInformers.Config().V1().Infrastructures().Lister(),
+		versionGetter:  versionGetter,
+		targetVersion:  targetVersion,
+		eventRecorder:  eventRecorder.WithComponentSuffix("CSIDriverStarter"),
+	}
+
+	// Populating all CSI driver operator ControllerManagers here simplifies
+	// the startup a lot
+	// - All necessary informers are populated.
+	// - CSO then just Run()s
+	// All ControllerManagers are not running at this point! They will be
+	// started in sync() when their platform is detected.
+	c.controllers = []csiDriverControllerManager{}
+	for _, cfg := range driverConfigs {
+		c.controllers = append(c.controllers, csiDriverControllerManager{
+			operatorConfig: cfg,
+			mgr:            c.createCSIControllerManager(cfg, clients, resyncInterval),
+			running:        false,
+		})
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(clients.OperatorClient).WithInformers(
+		clients.OperatorClient.Informer(),
+		clients.ConfigInformers.Config().V1().Infrastructures().Informer(),
+	).ToController("CSIDriverStarter", eventRecorder)
+}
+
+func (c *CSIDriverStarterController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	infrastructure, err := c.infraLister.Get(infraConfigName)
+	if err != nil {
+		return err
+	}
+
+	var platform configv1.PlatformType
+	if infrastructure.Status.PlatformStatus != nil {
+		platform = infrastructure.Status.PlatformStatus.Type
```

Comment:
Did you intend to use `csiDriverControllerManager.platform` instead?

I don't think you did, if so we can remove it

Reply:
removed `csiDriverControllerManager.platform`